src/kudu/consensus/routing.cc

```diff
@@ -507,6 +507,128 @@ string DurableRoutingTable::LogPrefix() const {
   return strings::Substitute("T $0 P $1: ", tablet_id_, fs_manager_->uuid());
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// SimpleRegionRoutingTable
+////////////////////////////////////////////////////////////////////////////////
+Status SimpleRegionRoutingTable::Create(
+    RaftConfigPB raft_config,
+    RaftPeerPB local_peer_pb,
+    std::shared_ptr<IRoutingTable>* srt) {
+
+  auto simple_routing_table = std::make_shared<SimpleRegionRoutingTable>();
+  simple_routing_table->SetLocalPeerPB(std::move(local_peer_pb));
+  simple_routing_table->UpdateRaftConfig(std::move(raft_config));
+  *srt = std::move(simple_routing_table);
+
+  return Status::OK();
+}
+
+Status SimpleRegionRoutingTable::RebuildProxyTopology(RaftConfigPB raft_config) {
+  ProxyTopologyPB proxy_topology;
+  const std::string& local_peer_region = local_peer_pb_.attrs().region();
+
+  // 1. Identify the 'proxy peer' for each region. The peer that is backed by a
+  // database in a region acts as a 'proxy peer' for the region. [Update when
+  // region splitting is supported]. If there are multiple such peers ina
+  // region, then pick the first peer as the 'proxy peer' for the region[this
+  // cannot happen once region splitting is supported]
+  // 2. Also build a map of "peer-uuid to peer-region" for all peers
+  std::unordered_map<std::string, std::string> region_proxy_peer_map;
+  std::unordered_map<std::string, std::string> peer_region_map;
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    if (peer.attrs().backing_db_present()) {
+      region_proxy_peer_map.emplace(
+          peer.attrs().region(), peer.permanent_uuid());
+    }
+    peer_region_map.emplace(peer.permanent_uuid(), peer.attrs().region());
+  }
+
+  // For every destination peer, choose the peer from which it will be proxied
+  // from. Some rules (see proxy_policy.h):
+  // 1. A peer with a backing database is never proxied. Leader ships messages
+  // directly to all such peers
+  // 2. A peer which is in the same region as the node that is shipping messages
+  // will not be proxied i.e all peers in the same region as the 'source' will
+  // get messages directly from the 'source'
+  // 3. A peer which is in a region without and valid 'proxy peer' will recieve
+  // messages directly from the 'source'
+  std::unordered_map<std::string, std::string> proxy_peer_map;
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    // Add a new edge into the topology
+    ProxyEdgePB* proxy_edge = proxy_topology.add_proxy_edges();
+    proxy_edge->set_peer_uuid(peer.permanent_uuid());
+
+    std::string peer_region = peer.attrs().region();
+    if (peer.attrs().backing_db_present()) {
+      // Peers that have a backing database are not proxied (rule #1)
+      proxy_edge->set_proxy_from_uuid(proxy_edge->peer_uuid());
```

Comment:
Why bother creating a self-loop edge? In the existing impl this would be `emplace(peer.uuid, leader.uuid)` which makes more sense to me (but as I am writing this comment I'm still reading through the impl trying to understand it).

Reply:
Agree. ProxyTopologyPB is not used in this simple routing policy and is maintained only for reporting purpose. Will see if I can change this